### PR TITLE
CAPT changes related to kubified tinkerbell stack

### PR DESCRIPTION
**Pull request by panktishah26**
## Description

This PR contains changes related to kubified tinkerbell stack. Tink PostgreSQL dependency is removed from the CAPT and CAPT controllers will only have to communicate with underlying Kubernetes client. 

Signed-off-by: Pankti Shah <pjshah@amazon.com>

## Why is this needed

We have kubified all the services of Tinkerbell stack like Tink-apis, Boots, Hegel, Rufio etc. We needed to change CAPT as well to make the whole kubified stack work to create a BareMetal cluster.

Fixes: #

## How Has This Been Tested?

To test these changes, 
1. I set up Kubified Tinkerbell stack by creating containers such as tink-server, boots, hegel, tink-controller, registry and k3s containing Kubernetes related changes. 
2. Removed PostgreSQL tink-api dependencies from the CAPT repo.
3. Accommodated all the tink-api v1alpha1 structure changes related to hardware, template and workflow.
4. Ran CAPT on k3s cluster using clusterctl api.
5. Fed hardware yaml files and deployment file to the running cluster which created a cluster with 1 control plane and 1 worker successfully. 

# Sample yaml file for new hardware CRDs

```
apiVersion: "tinkerbell.org/v1alpha1"
kind: Hardware
metadata:
  name: node-1
  namespace: default
spec:
  disks:
    - device: /dev/sda
  metadata:
    facility:
      facility_code: onprem
    instance:
      userdata: ""
      hostname: "node-1"
      id: "xx:xx:xx:xx:xx:xx"
      operating_system:
        distro: "ubuntu"
        os_slug: "ubuntu_20_04"
        version: "20.04"
  interfaces:
    - dhcp:
        arch: x86_64
        hostname: node-1
        ip:
          address: 0.0.0.0
          gateway: 0.0.0.1
          netmask: 255.255.255.0
        lease_time: 86400
        mac: xx:xx:xx:xx:xx:
        name_servers:
          - 8.8.8.8
        uefi: true
      netboot:
        allowPXE: true
        allowWorkflow: true
```

## How are existing users impacted? What migration steps/scripts do we need?
Existing users might have to take the latest versions of each Tinkerbell services to created a cluster.

## Checklist:

I have:

- [x] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

File: config/default/manager_image_patch.yaml [hunk 1]
```diff
@@ -8,5 +8,5 @@ spec:
     spec:
       containers:
       # Change the value of image field below to your controller image URL
-      - image: ghcr.io/tinkerbell/cluster-api-provider-tinkerbell:dev
+      - image: tinkerbell/cluster-api-provider-tinkerbell-amd64:dev
```
Comment (micahhausler) [inline config/default/manager_image_patch.yaml:11]: We'll build a multi-arch image, so this can probably just be
```suggestion
      - image: tinkerbell/cluster-api-provider-tinkerbell:dev
```

File: controllers/machine.go [hunk 2]
```diff
@@ -326,6 +328,27 @@ func (mrc *machineReconcileContext) ensureHardwareUserData(hardware *tinkv1.Hard
...
+       if metadata != nil {
+			instances := metadata.Instance
+			if instances != nil {
+					storage := metadata.Instance.Storage
+					if storage != nil {
+							disks := metadata.Instance.Storage.Disks
+							diskLen := len(disks)
+							for i := 0; i < diskLen; i++ {
+									foundDisks = append(foundDisks, tinkv1.Disk{Device: disks[i].Device})
+							}
```
Comment (micahhausler) [inline controllers/machine.go:343]: you can `range` over the disks here
```suggestion
							for _, disk := range metadata.Instance.Storage.Disks {
									foundDisks = append(foundDisks, tinkv1.Disk{Device: disk.Device})
							}
```

File: controllers/machine.go [hunk 3]
```diff
@@ -402,7 +432,7 @@ func (mrc *machineReconcileContext) createWorkflow() error {
 		},
 		Spec: tinkv1.WorkflowSpec{
 			TemplateRef: mrc.tinkerbellMachine.Name,
-			HardwareRef: mrc.tinkerbellMachine.Spec.HardwareName,
+			HardwareMap: map[string]string{"device_1": ""},
```
Comment (micahhausler) [inline controllers/machine.go:435]: You'll need to reference the hardware name in the workflow

File: config/default/manager_pull_policy.yaml [hunk 4]
```diff
@@ -8,4 +8,4 @@ spec:
     spec:
       containers:
       - name: manager
-        imagePullPolicy: IfNotPresent
+        imagePullPolicy: Always
```
Comment (micahhausler) [inline config/default/manager_pull_policy.yaml:11]: I think we'll want to keep `IfNotPresent` in the repo, but I'm guessing this is just for development?

File: controllers/machine.go [hunk 5]
```diff
@@ -326,6 +328,27 @@ func (mrc *machineReconcileContext) ensureHardwareUserData(hardware *tinkv1.Hard
...
 }
 
+func (mrc *machineReconcileContext) ensureHardwareDiskData(hardware *tinkv1.Hardware) error {
+       foundDisks := make([]tinkv1.Disk, 0)
+       metadata := hardware.Spec.Metadata
+       if metadata != nil {
+			instances := metadata.Instance
+			if instances != nil {
+					storage := metadata.Instance.Storage
+					if storage != nil {
```
Comment (micahhausler) [inline controllers/machine.go:338]: I think you can make this a single `if` block
```suggestion
       if hardware.Spec.Metadata != nil && hardware.Spec.Metadata.Instance  != nil && hardware.Spec.Metadata.Instance.Storage  != nil {
```

File: controllers/machine.go [hunk 6]
```diff
@@ -388,9 +416,11 @@ func (mrc *machineReconcileContext) workflowExists() (bool, error) {
 }
 
 func (mrc *machineReconcileContext) createWorkflow() error {
+
```
Comment (micahhausler) [inline controllers/machine.go:419]: nit: can you remove whitespace-only changes?

**Review by micahhausler — CHANGES_REQUESTED** [review]
Thanks for putting out what you've got so far! I have just a few suggestions

File: controllers/machine.go [hunk 7]
```diff
@@ -330,6 +336,26 @@ func (mrc *machineReconcileContext) ensureHardwareUserData(hardware *tinkv1.Hard
...
+			}
+		} else {
+			return ErrPatchingHardwareDiskData
+		}
+
+		hardware.Spec.Disks = foundDisks
+	}
+
+	return nil
+}
```
Comment (abhinavmpandey08) [inline controllers/machine.go:357]: Why is this needed here? This seems like something that should be handled by tink-controller and not CAPT?

File: controllers/tinkerbellmachine_controller.go [hunk 8]
```diff
@@ -45,6 +45,9 @@ type TinkerbellMachineReconciler struct {
 // +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=tinkerbellmachines/status,verbs=get;update;patch
 // +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machines;machines/status,verbs=get;list;watch
 // +kubebuilder:rbac:groups="",resources=secrets;,verbs=get;list;watch
+// +kubebuilder:rbac:groups=tinkerbell.org,resources=hardware;hardware/status,verbs=get;list;watch;create;update;patch;delete
```
Comment (micahhausler) [inline controllers/tinkerbellmachine_controller.go:48]: You can drop the create and delete permissions here

**Review by micahhausler — CHANGES_REQUESTED** [review]

File: config/default/manager_image_patch.yaml [hunk 9]
```diff
@@ -8,5 +8,5 @@ spec:
     spec:
       containers:
       # Change the value of image field below to your controller image URL
-      - image: ghcr.io/tinkerbell/cluster-api-provider-tinkerbell:dev
+      - image: ghcr.io/tinkerbell/cluster-api-provider-tinkerbell-amd64:dev
```
Comment (micahhausler) [inline config/default/manager_image_patch.yaml:11]: Can we drop the `-amd64` again?

File: config/default/manager_pull_policy.yaml [hunk 10]
```diff
@@ -8,4 +8,4 @@ spec:
     spec:
       containers:
       - name: manager
-        imagePullPolicy: IfNotPresent
```
Comment (micahhausler) [inline config/default/manager_pull_policy.yaml:11]: Can we change back to `IfNotPresent`?

File: config/default/manager_image_patch.yaml (hunk 9 above)
Comment (ptrivedi) [reply config/default/manager_image_patch.yaml:11]: @pjshah, when you build capt locally using "make docker-build", this file as well as the manager_pull_policy.yaml files below are updated automatically with these changes. You will need to revert these back.

File: config/default/manager_pull_policy.yaml (hunk 10 above)
Comment (ptrivedi) [reply config/default/manager_pull_policy.yaml:11]: Refer to comment above, @panktishah26

**Review by micahhausler — APPROVED** [review]
LGTM! Great work @panktishah26!

File: docs/README.md [hunk 11]
```diff
@@ -65,11 +65,6 @@ cat <<EOF > tilt-settings.json
   "provider_repos": ["../cluster-api-provider-tinkerbell"],
   "enable_providers": ["tinkerbell", "kubeadm-bootstrap", "kubeadm-control-plane"],
   "allowed_contexts": ["<your kubeconfig context to use"],
```
Comment (micahhausler) [inline docs/README.md:67]: Can you remove the trailing comma to make it valid JSON?

File: docs/QUICK-START.md [hunk 12]
```diff
@@ -239,22 +175,52 @@ management cluster.
...
+    - metadata.spec.disks is set and contains at least one device matching an available disk on the system
+  - An example of a valid hardware definition for use with the Tinkerbell provider:
 
-**NOTE:** The id should match the id of the Tinkerbell Hardware resource and the name will need to be unique.
+    ```yaml
+    apiVersion: "tinkerbell.org/v1alpha1"
+    kind: Hardware
+    metadata:
+      name: node-1
+      namespace: default
```
Comment (micahhausler) [inline docs/QUICK-START.md:190]: Can you add the labels above to this example?

**Review by micahhausler — CHANGES_REQUESTED** [review]
Thanks for updating the docs! Just some minor changes

**Review by micahhausler — APPROVED** [review]
LGTM!

File: controllers/machine.go [hunk 13]
```diff
@@ -345,7 +349,7 @@ func (mrc *machineReconcileContext) ensureHardware() (*tinkv1.Hardware, error) {
 	}
 
 	mrc.tinkerbellMachine.Spec.HardwareName = hardware.Name
-	mrc.tinkerbellMachine.Spec.ProviderID = fmt.Sprintf("tinkerbell://%s", hardware.Spec.ID)
+	mrc.tinkerbellMachine.Spec.ProviderID = fmt.Sprintf("tinkerbell://%s", hardware.UID)
```
Comment (detiber) [inline controllers/machine.go:352]: Use of the UID here would cause issues if types are ever backed up and restored (or if resources are "pivoted" to a different Kubernetes cluster). This wasn't an issue with the Tinkerbell when using the Postgres backend because users because users were specifying the UUID when creating the hardware.

Before recommending an alternative approach, I have a few questions to narrow down what supported permutations we expect to exist.

- Would we expect to be able to reference hardware defined in a different namespace to be used for a given Tinkerbell Machine?
- Would we expect the use of hardware defined on a different k8s cluster to be used?
- Would we expect to be able to reference hardware defined using Tinkerbell with the postgresql backend still?

File: controllers/machine.go [hunk 14]
```diff
@@ -501,7 +507,7 @@ func (mrc *machineReconcileContext) createWorkflow() error {
 		},
 		Spec: tinkv1.WorkflowSpec{
 			TemplateRef: mrc.tinkerbellMachine.Name,
-			HardwareRef: mrc.tinkerbellMachine.Spec.HardwareName,
+			HardwareMap: map[string]string{"device_1": hardware.Spec.Metadata.Instance.ID},
```
[the remark this answers is not in the document]
Comment (detiber) [reply controllers/machine.go:510]: Is the assumption here that hardware.Spec.Metadata.Instance.ID is going to match the MAC address?

I'm wondering if there is a better way to handle this mapping that provides a clearer mapping between Hardware and Workflow.

Maybe something like this (simplified yaml):
```
hardwareRefs:
- placeholder: "device_1",
- hardwareName: mrc.tinkerbellMachine.Spec.HardwareName,
<some way to specify interface index and whether to use IP or MAC for matching>
```

This would potentially give the same flexibility as the current map without losing an easy way to match the hardware associated with a given workflow (especially for the purposes in displaying in `kubectl get`)

**Review by detiber — COMMENTED** [review]
Love seeing this work coming together, just a few questions and suggestions. Please excuse my ignorance if anything I brought up has already been discussed while I've been offline.

File: controllers/machine.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (micahhausler) [reply controllers/machine.go:510]: > Still reading through the changes, but initial reaction here is: should this be a well defined data type rather than a map[string]string here?

 I think this could use a better data type, but that will necessitate changes to the API now defined in Tinkerbell.  

> Is the assumption here that hardware.Spec.Metadata.Instance.ID is going to match the MAC address?

Yes. We should come up with a better way to encode this, but its probably outside the scope of this PR

File: go.mod [hunk 15]
```diff
@@ -5,16 +5,14 @@ go 1.17
 require (
 	github.com/go-logr/logr v0.4.0
 	github.com/google/uuid v1.3.0
-	github.com/onsi/gomega v1.19.0
+	github.com/onsi/gomega v1.17.0
```
[the remark this answers is not in the document]
Comment (panktishah26) [reply go.mod:8]: I guess it got downgraded when I ran go mod tidy. Let me revert this change. Thanks for pointing that out.

File: go.mod [hunk 16]
```diff
@@ -5,16 +5,14 @@ go 1.17
...
-	google.golang.org/protobuf v1.28.0
-	k8s.io/api v0.22.9
-	k8s.io/apimachinery v0.22.9
-	k8s.io/client-go v0.22.9
-	k8s.io/component-base v0.22.9
+	github.com/tinkerbell/tink v0.6.1-0.20220419201134-2b877141d755
+	k8s.io/api v0.22.8
+	k8s.io/apimachinery v0.22.8
+	k8s.io/client-go v0.22.8
+	k8s.io/component-base v0.22.8
```
[the remark this answers is not in the document]
Comment (panktishah26) [reply go.mod:15]: Same as above. I will revert this.

File: controllers/machine.go (hunk 13 above)
Comment (panktishah26) [reply controllers/machine.go:352]: Should we make this change in this PR or should I create an another PR?

File: controllers/machine.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (detiber) [reply controllers/machine.go:510]: Cleaning this up in the future sounds good to me

File: controllers/machine.go (hunk 13 above)
Comment (micahhausler) [reply controllers/machine.go:352]: @panktishah26 its up to you

Comment (detiber) [reply controllers/machine.go:352]: @micahhausler thanks for the clarification, it really helps.

I don't think we need to. block the PR on updating the providerID, especially since it would require additional changes to the controllers to support the new pattern.

I would consider getting things updated to use `tinkerbell//{namespace}/{name}` a release blocker, though to avoid users hitting issues related to backup/restore and pivot.

**Comment by panktishah26** [comment]
I have reverted go.mod file change. Kindly let me know if there is any other changes needs to be done. @micahhausler @detiber

**Review by micahhausler — APPROVED** [review]
LGTM, awesome work @panktishah26!
